src/core/lib/slice/slice_buffer.cc

```diff
@@ -370,6 +370,24 @@ grpc_slice grpc_slice_buffer_take_first(grpc_slice_buffer* sb) {
   return slice;
 }
 
+void grpc_slice_buffer_consume_first(grpc_slice_buffer* sb) {
+  GPR_ASSERT(sb->count > 0);
+  sb->length -= GRPC_SLICE_LENGTH(sb->slices[0]);
+  grpc_slice_unref_internal(sb->slices[0]);
+  sb->slices++;
+  if (--sb->count == 0) {
+    sb->slices = sb->base_slices;
+  }
+}
+
+void grpc_slice_buffer_sub_first(grpc_slice_buffer* sb, size_t begin,
+                                 size_t end) {
+  // TODO(soheil): Introduce a ptr version for sub.
+  sb->length -= GRPC_SLICE_LENGTH(sb->slices[0]);
+  sb->slices[0] = grpc_slice_sub_no_ref(sb->slices[0], begin, end);
+  sb->length += GRPC_SLICE_LENGTH(sb->slices[0]);
```

Comment:
I'd prefer length versus (end-begin) mainly for readability, and performance shouldn't be different. Once I introduced the ptr version for `grpc_slice_sub`, this will go away.

Reply:
Doesn't GRPC_SLICE_LENGTH impose an additional branch, since we have to look up whether it's an inlined slice or not? (And based on that, we either convert a uint8_t or just read a size_t).

Reply:
Yes, it's a cmove, but I don't think it matters here. But sure changed to end - start because I think you feel strongly about that. 

The bottleneck here is the TODO above, addressing that dwarfs the overheads here.